File: src/numeric/max/max.hs
{-
    Get the maximum value of some sequence of values
 -}
import System.Environment (getArgs)
import Data.List (intersperse)
import Data.Maybe (fromMaybe)
import Text.Read (readMaybe)

-- | Get the maximum value of some list
max_list :: Ord a => [a] -> Maybe a
max_list [] = Nothing
max_list (x : xs) = Just (reduce_max xs x)
    where
        reduce_max [] max = max
        reduce_max (x : xs) max
            | x > max = reduce_max xs x
            | otherwise = reduce_max xs max

-- | Get the arguments as integers
argsAsInts :: [String] -> (Maybe [Integer])
argsAsInts [] = Nothing
argsAsInts args = traverse readMaybe args

-- | The main entry point to the application
main :: IO ()
main = do
    args <- getArgs
    let intArgs = fromMaybe [15, 10] (argsAsInts args)
    let maxValue = max_list intArgs

    putStrLn $ "values: [" ++ (concat (intersperse " " (map show intArgs))) ++ "]"
    putStrLn $ "max: " ++ case maxValue of
        Nothing -> "n/a"
        Just value -> show value

File: src/numeric/max/max.scala
/*
  Get the maximum value of some sequence of values
 */
import scala.annotation.tailrec
import scala.math.Ordering

/**
 * Get the maximum value of some list
 *
 * @param list The list of values to find the maximum of
 * @tparam T The type of the values in the list
 * @return An Option containing the maximum value, or None if the list is empty
 */
def max_list[T: Ordering](list: List[T]): Option[T] =
  @tailrec
  def reduce_max(list: List[T], prior: T): T =
    list match {
      case Nil => prior
      case head :: tail =>
          reduce_max(tail,
            if (implicitly[Ordering[T]].gt(head, prior)) then
              head
            else prior)
    }

  list match
    case Nil => None
    case head :: tail => Some(reduce_max(tail, head))

/**
 * The main entry point to the application
 */
@main
def main(args: String*): Unit =
  val list = if (args.size > 0) then args.map(_.toInt).toList else List(15, 10)
  val max = max_list(list)
  println(f"values: $list")
  println(f"max: ${max.getOrElse("n/a")}")
  